File: lib/services/meal_template_generator.dart
import 'package:flutter/foundation.dart';
import 'package:uuid/uuid.dart';

import '../core/app_language.dart';
import '../core/template_origin.dart';
import '../data/catalog/starter_foods.dart';
import '../data/db/drift_database.dart';
import '../features/meals/data/repositories.dart';
import 'meal_portion_solver.dart';
import 'meal_recipes.dart';
import 'profile_fit.dart';
import 'user_profile_service.dart';

/// Builds meal templates for a profile by **selecting from the tagged food
/// catalog**, sized against the user's calorie and protein targets.
///
/// The previous version did two things this one deliberately does not:
///
///  * It inserted its own foods, hardcoded per diet type, duplicating what
///    the catalog already held.
///  * It decided what a food contained by matching its English name against
///    literal lists (`['Greek Yogurt','Cottage Cheese','Milk','Butter']`),
///    so any food the user added was invisible to the filter, Hebrew names
///    never matched, and adding a food to the catalog silently broke it.
///
/// Both are now handled by [FoodTag] + [ProfileFit], which the browsing UI
/// uses too -- so what gets generated and what gets shown can't disagree.
class MealTemplateGenerator {
  MealTemplateGenerator(this._database, this._profile, this._language);

  final AppDatabase _database;
  final UserProfile _profile;

  /// The language every template this generator writes is named in. Resolved
  /// once, here -- see the note on `WorkoutTemplateGenerator._language`.
  final AppLanguage _language;

  final _uuid = const Uuid();

  Future<List<MealTemplateData>> generateTemplates() async {
    final available = await _availableFoods();
    if (available.isEmpty) {
      debugPrint('[MEAL-GEN] No foods fit this profile; skipping');
      return const [];
    }
    final byRecipeKey = _byRecipeKey(available);

    final meals = _mealPlan();
    final created = <MealTemplateData>[];
    final usedRecipes = <String>{};

    for (var i = 0; i < meals.length; i++) {
      final meal = meals[i];
      final recipe = _pickRecipe(meal.kind, byRecipeKey, usedRecipes);
      if (recipe == null) continue;
      usedRecipes.add(recipe.name);

      final resolved = _resolve(recipe, byRecipeKey);
      final portions = MealPortionSolver.solve(
        protein: resolved[RecipeRole.protein],
        carb: resolved[RecipeRole.carb],
        fat: resolved[RecipeRole.fat],
        veg: resolved[RecipeRole.produce],
        // The lean protein is what lets a dish hold its protein target while
        // calories come down -- egg whites alongside whole eggs.
        extras: [
          if (resolved[RecipeRole.leanProtein] != null)
            resolved[RecipeRole.leanProtein]!,
        ],
        kcalTarget: _profile.calorieTarget * meal.fraction,
        proteinTarget: _profile.proteinTargetG * meal.fraction,
        carbsTarget: _profile.carbsTargetG * meal.fraction,
        fatTarget: _profile.fatTargetG * meal.fraction,
      );
      if (portions.isEmpty) continue;

      final now = DateTime.now();
      final template = MealTemplateData(
        id: _uuid.v4(),
        // Named for the dish, prefixed with when it is eaten, so the list
        // reads like a meal plan rather than a set of macro buckets. Composed
        // in one language: an untranslated recipe falls back to English for
        // the dish *and* the slot, so a title is never half-Hebrew.
        name: _language == AppLanguage.hebrew && recipe.nameHe != null
            ? '${meal.nameHe}: ${recipe.nameHe}'
            : '${meal.name}: ${recipe.name}',
        description: _language == AppLanguage.hebrew
            ? (recipe.descriptionHe ?? recipe.description)
            : recipe.description,
        origin: TemplateOrigin.generated,
        createdAt: now,
        updatedAt: now,
      );
      await _database.insertMealTemplate(template);

      for (final portion in portions) {
        await _database.insertMealTemplateItem(MealTemplateItemData(
          id: _uuid.v4(),
          templateId: template.id,
          foodId: portion.food.id,
          amount: portion.amount,
        ));
      }
      created.add(template);
    }

    debugPrint('[MEAL-GEN] Generated ${created.length} meal templates');
    return created;
  }

  /// The available foods, keyed by the **English catalog name** a recipe
  /// names them by.
  ///
  /// Recipes name their ingredients in English ('Eggs', 'Whole Wheat Bread')
  /// because that is what a recipe is about, and those names are stable. The
  /// seeded row's own `name` is not a usable key any more: it is written in
  /// whichever language the user chose, so on a Hebrew install every recipe
  /// lookup missed and the generator produced **zero** meal templates while
  /// reporting nothing wrong.
  ///
  /// Going through the starter catalog's id restores a language-independent
  /// join: recipe name -> starter id -> the seeded row, whatever it is called.
  /// A food the user added themselves is not reachable this way, which is
  /// correct -- recipes are written against the shipped catalog.
  static Map<String, FoodItemData> _byRecipeKey(List<FoodItemData> available) {
    final byId = {for (final f in available) f.id: f};
    return {
      for (final starter in StarterFoodCatalog.all)
        if (byId[starter.id] != null) starter.name: byId[starter.id]!,
    };
  }

  /// The first recipe for this slot whose required ingredients all exist and
  /// suit the profile, preferring one not already used today so a day is not
  /// the same dish three times.
  MealRecipe? _pickRecipe(
    MealSlotKind kind,
    Map<String, FoodItemData> byName,
    Set<String> used,
  ) {
    final candidates = MealRecipes.forKind(kind);
    MealRecipe? fallback;
    for (final recipe in candidates) {
      if (!_canMake(recipe, byName)) continue;
      fallback ??= recipe;
      if (!used.contains(recipe.name)) return recipe;
    }
    // Every makeable recipe already used -- repeat rather than skip a meal.
    return fallback;
  }

  bool _canMake(MealRecipe recipe, Map<String, FoodItemData> byName) =>
      recipe.slots.every(
          (slot) => !slot.required || slot.candidates.any(byName.containsKey));

  /// Resolves each slot to the first candidate food the profile allows.
  Map<RecipeRole, FoodItemData> _resolve(
    MealRecipe recipe,
    Map<String, FoodItemData> byName,
  ) {
    final resolved = <RecipeRole, FoodItemData>{};
    for (final slot in recipe.slots) {
      if (resolved.containsKey(slot.role)) continue;
      for (final name in slot.candidates) {
        final food = byName[name];
        if (food != null) {
          resolved[slot.role] = food;
          break;
        }
      }
    }
    return resolved;
  }

  Future<List<FoodItemData>> _availableFoods() async {
    final all = await _database.getAllFoods();
    return all
        .where((data) => ProfileFit.foodFits(foodItemFromData(data), _profile))
        .toList();
  }

  /// How the day's calories split across meals, and what each is called.
  List<_MealSlot> _mealPlan() {
    switch (_profile.mealCountPerDay) {
      case '2':
        return const [
          _MealSlot('Brunch', MealSlotKind.breakfast, 0.45, 'בראנץ׳'),
          _MealSlot('Dinner', MealSlotKind.main, 0.55, 'ארוחת ערב'),
        ];
      case '4':
        return const [
          _MealSlot('Breakfast', MealSlotKind.breakfast, 0.25, 'ארוחת בוקר'),
          _MealSlot('Lunch', MealSlotKind.main, 0.30, 'ארוחת צהריים'),
          _MealSlot('Snack', MealSlotKind.snack, 0.20, 'חטיף'),
          _MealSlot('Dinner', MealSlotKind.main, 0.25, 'ארוחת ערב'),
        ];
      case 'intermittent_fasting_16_8':
        return const [
          _MealSlot('First Meal', MealSlotKind.breakfast, 0.40, 'ארוחה ראשונה'),
          _MealSlot('Second Meal', MealSlotKind.main, 0.35, 'ארוחה שנייה'),
          _MealSlot('Final Meal', MealSlotKind.main, 0.25, 'ארוחה אחרונה'),
        ];
      case '3':
      default:
        return const [
          _MealSlot('Breakfast', MealSlotKind.breakfast, 0.30, 'ארוחת בוקר'),
          _MealSlot('Lunch', MealSlotKind.main, 0.40, 'ארוחת צהריים'),
          _MealSlot('Dinner', MealSlotKind.main, 0.30, 'ארוחת ערב'),
        ];
    }
  }
}

class _MealSlot {
  const _MealSlot(this.name, this.kind, this.fraction, this.nameHe);
  final String name;

  /// Hebrew label for the slot ("Breakfast" -> "ארוחת בוקר"), used to compose
  /// the template's `nameHe`.
  final String nameHe;
  final MealSlotKind kind;

  /// Share of the day's calorie and protein targets this meal carries.
  final double fraction;
}

File: Packages/WellnessKit/Sources/WellnessDomain/MealTemplateGenerator.swift
import Foundation
import WellnessModels

public struct MealTemplateGenerator: Sendable {
    public let profile: UserProfile
    public let language: AppLanguage

    public init(profile: UserProfile, language: AppLanguage) {
        self.profile = profile
        self.language = language
    }

    public struct GeneratedMealTemplate: Sendable {
        public let template: MealTemplate
        public let items: [MealTemplateItem]
    }

    /// - Parameter catalogNameById: Maps food id -> English catalog name (from StarterFood).
    ///   Recipes reference foods by English name; this mapping lets resolution work
    ///   regardless of what language the seeded row is stored in.
    public func generateTemplates(
        availableFoods: [FoodItem],
        catalogNameById: [String: String],
        makeId: () -> String
    ) -> [GeneratedMealTemplate] {
        let fitting = availableFoods.filter { ProfileFit.foodFits($0, profile) }
        if fitting.isEmpty { return [] }

        let byRecipeKey = Self.buildRecipeKeyMap(fitting, catalogNameById: catalogNameById)
        let meals = mealPlan()
        var created: [GeneratedMealTemplate] = []
        var usedRecipes = Set<String>()

        for meal in meals {
            guard let recipe = pickRecipe(meal.kind, byName: byRecipeKey, used: usedRecipes) else {
                continue
            }
            usedRecipes.insert(recipe.name)

            let resolved = resolve(recipe, byName: byRecipeKey)
            let portions = MealPortionSolver.solve(
                protein: resolved[.protein],
                carb: resolved[.carb],
                fat: resolved[.fat],
                veg: resolved[.produce],
                extras: resolved[.leanProtein].map { [$0] } ?? [],
                kcalTarget: profile.calorieTarget * meal.fraction,
                proteinTarget: profile.proteinTargetG * meal.fraction,
                carbsTarget: profile.carbsTargetG * meal.fraction,
                fatTarget: profile.fatTargetG * meal.fraction
            )
            if portions.isEmpty { continue }

            let templateId = makeId()
            let name: String
            if language == .hebrew, let hebrewName = recipe.nameHe {
                name = "\(meal.nameHe): \(hebrewName)"
            } else {
                name = "\(meal.name): \(recipe.name)"
            }

            let template = MealTemplate(
                id: templateId,
                name: name,
                description: language == .hebrew
                    ? (recipe.descriptionHe ?? recipe.description)
                    : recipe.description,
                origin: .generated
            )

            let items = portions.map { portion in
                MealTemplateItem(
                    id: makeId(),
                    templateId: templateId,
                    foodId: portion.food.id,
                    amount: portion.amount
                )
            }

            created.append(GeneratedMealTemplate(template: template, items: items))
        }
        return created
    }

    // MARK: - Recipe key map

    static func buildRecipeKeyMap(
        _ available: [FoodItem],
        catalogNameById: [String: String]
    ) -> [String: FoodItem] {
        let byId = Dictionary(available.map { ($0.id, $0) }, uniquingKeysWith: { _, b in b })
        var out: [String: FoodItem] = [:]
        for (id, englishName) in catalogNameById {
            if let food = byId[id] { out[englishName] = food }
        }
        return out
    }

    // MARK: - Recipe picking

    private func pickRecipe(
        _ kind: MealSlotKind,
        byName: [String: FoodItem],
        used: Set<String>
    ) -> MealRecipe? {
        let candidates = MealRecipes.forKind(kind)
        var fallback: MealRecipe?
        for recipe in candidates {
            guard canMake(recipe, byName: byName) else { continue }
            if fallback == nil { fallback = recipe }
            if !used.contains(recipe.name) { return recipe }
        }
        return fallback
    }

    private func canMake(_ recipe: MealRecipe, byName: [String: FoodItem]) -> Bool {
        recipe.slots.allSatisfy { slot in
            !slot.required || slot.candidates.contains(where: { byName[$0] != nil })
        }
    }

    private func resolve(
        _ recipe: MealRecipe,
        byName: [String: FoodItem]
    ) -> [RecipeRole: FoodItem] {
        var resolved: [RecipeRole: FoodItem] = [:]
        for slot in recipe.slots {
            if resolved[slot.role] != nil { continue }
            for name in slot.candidates {
                if let food = byName[name] {
                    resolved[slot.role] = food
                    break
                }
            }
        }
        return resolved
    }

    // MARK: - Meal plan

    private struct MealSlot {
        let name: String
        let nameHe: String
        let kind: MealSlotKind
        let fraction: Double
    }

    private func mealPlan() -> [MealSlot] {
        switch profile.mealCountPerDay {
        case "2":
            return [
                MealSlot(name: "Brunch", nameHe: "בראנץ׳", kind: .breakfast, fraction: 0.45),
                MealSlot(name: "Dinner", nameHe: "ארוחת ערב", kind: .main, fraction: 0.55),
            ]
        case "4":
            return [
                MealSlot(name: "Breakfast", nameHe: "ארוחת בוקר", kind: .breakfast, fraction: 0.25),
                MealSlot(name: "Lunch", nameHe: "ארוחת צהריים", kind: .main, fraction: 0.30),
                MealSlot(name: "Snack", nameHe: "חטיף", kind: .snack, fraction: 0.20),
                MealSlot(name: "Dinner", nameHe: "ארוחת ערב", kind: .main, fraction: 0.25),
            ]
        case "intermittent_fasting_16_8":
            return [
                MealSlot(name: "First Meal", nameHe: "ארוחה ראשונה", kind: .breakfast, fraction: 0.40),
                MealSlot(name: "Second Meal", nameHe: "ארוחה שנייה", kind: .main, fraction: 0.35),
                MealSlot(name: "Final Meal", nameHe: "ארוחה אחרונה", kind: .main, fraction: 0.25),
            ]
        default:
            return [
                MealSlot(name: "Breakfast", nameHe: "ארוחת בוקר", kind: .breakfast, fraction: 0.30),
                MealSlot(name: "Lunch", nameHe: "ארוחת צהריים", kind: .main, fraction: 0.40),
                MealSlot(name: "Dinner", nameHe: "ארוחת ערב", kind: .main, fraction: 0.30),
            ]
        }
    }
}
